library/ssl_tls13_server.c

```diff
@@ -22,18 +22,160 @@
 #if defined(MBEDTLS_SSL_SRV_C) && defined(MBEDTLS_SSL_PROTO_TLS1_3)
 
 #include "mbedtls/debug.h"
+#include "mbedtls/error.h"
 
 #include "ssl_misc.h"
 #include "ssl_debug_helpers.h"
 
+
+#if defined(MBEDTLS_KEY_EXCHANGE_WITH_CERT_ENABLED)
+#define SSL_CERTIFICATE_REQUEST_SEND_REQUEST 0
+#define SSL_CERTIFICATE_REQUEST_SKIP         1
+/* Coordination:
+ * Check whether a CertificateRequest message should be written.
+ * Returns a negative code on failure, or
+ * - SSL_CERTIFICATE_REQUEST_SEND_REQUEST
+ * - SSL_CERTIFICATE_REQUEST_SKIP
+ * indicating if the writing of the CertificateRequest
+ * should be skipped or not.
+ */
+static int ssl_tls13_certificate_request_coordinate( mbedtls_ssl_context *ssl )
+{
+    int authmode;
+
+    if( mbedtls_ssl_tls13_some_psk_enabled( ssl ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "<= skip write certificate request" ) );
+        return( SSL_CERTIFICATE_REQUEST_SKIP );
+    }
+#if defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    if( ssl->handshake->sni_authmode != MBEDTLS_SSL_VERIFY_UNSET )
+        authmode = ssl->handshake->sni_authmode;
+    else
+#endif /* MBEDTLS_SSL_SERVER_NAME_INDICATION */
+        authmode = ssl->conf->authmode;
```

Comment:
Is the placement of the `#endif` intended? 

Considering the the following line is checking if `uthmode == MBEDTLS_SSL_VERIFY_NONE` ?

Is ssl->conf->authmode defaulting to `MBEDTLS_SSL_VERIFY_NONE` ?

I am mostly wondering under the context that `MBEDTLS_SSL_SERVER_NAME_INDICATION` is undefined.

Reply:
Yes, the placement of the #endif is intended.
ssl->conf->authmode is default to MBEDTLS_SSL_VERIFY_NONE, but it can be set by the command of server by the function mbedtls_ssl_conf_authmode().